python4j/python4j-core/src/main/java/org/eclipse/python4j/PythonGC.java

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Konduit K.K.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+
+package org.eclipse.python4j;
+
+import org.bytedeco.cpython.PyObject;
+
+import java.io.Closeable;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.bytedeco.cpython.global.python.*;
+
+public class PythonGC implements Closeable{
+
+    private PythonGC previousFrame = null;
+    private boolean active = true;
+    private static PythonGC currentFrame = new PythonGC();
+
+    private Set<PyObject> objects = new HashSet<>();
+
+    private boolean alreadyRegistered(PyObject pyObject){
+        if (objects.contains(pyObject)){
+            return true;
+        }
+        if (previousFrame == null){
+            return false;
+        }
+        return previousFrame.alreadyRegistered(pyObject);
+
+    }
+
+    private void addObject(PythonObject pythonObject){
+        if (!active)return;
+        if (pythonObject.toString().equals("<NULL>")) return;
+        if (alreadyRegistered(pythonObject.getNativePythonObject())){
+            return;
+        }
+        objects.add(pythonObject.getNativePythonObject());
+    }
+
+    public static void register(PythonObject pythonObject){
+        currentFrame.addObject(pythonObject);
+    }
+
+    public static void keep(PythonObject pythonObject){
+        currentFrame.objects.remove(pythonObject.getNativePythonObject());
+        if (currentFrame.previousFrame != null){
+            currentFrame.previousFrame.addObject(pythonObject);
+        }
+    }
+
+    private PythonGC(){}
+
+    public static  PythonGC watch(){
+        PythonGC ret = new PythonGC();
+        ret.previousFrame = currentFrame;
+        ret.active = currentFrame.active;
+        currentFrame = ret;
+        return ret;
+    }
```

Comment:
Still can be issues with multiple threads calling watch simultaneously?

Reply:
```java
try (PythonGIL gil = PythonGIL.lock()){
    try(PythonGC gc = PythonGC.watch()){
        .......
    }
}
```